lom/src/pluginmgr/pluginmgr_common/pluginmgr_helper.go

```diff
@@ -115,63 +166,449 @@ func (plmgr *PluginManager) getPluginMetadata(plgname string) (plugins_common.IP
-                        } else {
-                            lomcommon.LogInfo("In RecvServerRequest: Received action request for plugin %v", plugin)
-
-                            /* TODO: Goutham : Handle Request, Do error checks, pass HeartBEat channel, handle timeouts, handle heartbeats etc
-                               lomcommon.GetGoroutineTracker().Start("plg_mgr_Run_Action_"+actionReq.Action+"_"+lomcommon.GetUUID(),
-                                   func() {
-                                       hbchan := make(chan plugins_common.PluginHeartBeat)// TODO: Goutham ; Do we need buffered instead ??
-                                       res := plugin.Request(hbchan, actionReq)
-                                       plmgr.clientTx.SendServerResponse(&lomipc.MsgSendServerResponse{
-                                           ReqType: lomipc.TypeServerRequestAction,
-                                           ResData: res,
-                                       })
-                                   })
-                            */
-                        }
+                        plmgr.setShutdownStatus(true)
+                        // start with native go routine only to handle goroutine tracker cleaneup gracefully on shutdown without deadlocks
+                        go plmgr.handleShutdown()
                     }
-                case lomipc.TypeServerRequestShutdown:
-                    // TODO: Goutham :  handle shutdown(wait for timeout), haqndle synchronously, do not listen on responses from plugins, send deregister with server/
-                    // exit process, use goroutinrtracker to see if any routines are running and log to syslog
-                default:
-                    lomcommon.LogError("RecvServerRequest() : Unknown server request type : %v", serverReq.ReqType)
                 }
+            default:
+                lomcommon.LogError("In run() RecvServerRequest : Unknown server request type : %v", serverReq.ReqType)
             }
         }
     }
     return nil
 }
 
-// TODO: Goutham : For all plugins call shutdown, change plugin state, wait for a timeout, Log error on any running go routines per tracker, etc
-func (plmgr *PluginManager) shutdown() error {
-    if err := plmgr.clientTx.DeregisterClient(); err != nil {
-        lomcommon.LogError("Error in deregistering client")
+/*
+ * verify request message
+ */
+func (plmgr *PluginManager) verifyRequestMsg(actionReq *lomipc.ActionRequestData) (plugins_common.Plugin, plugins_common.IPluginMetadata, error) {
+    if actionReq == nil {
+        return nil, nil, lomcommon.LogError("verifyRequestMsg() : nil request data")
+    }
+
+    if actionReq.Action == "" {
+        return nil, nil, lomcommon.LogError("verifyRequestMsg() : empty action")
+    }
+
+    // anamolykey used to track plugin repsonses for a given action for misbehaving plugins
+    if actionReq.AnomalyKey == "" {
```

Comment:
How is this handled as it is expected to be empty for detection requests?

Reply:
Yes you are  correct. For detection AnomalyKey is empty. I removed the check altogether as AnomalyKey  is only used from responses from plugin.